src/zcash/Proof.cpp

```diff
@@ -0,0 +1,185 @@
+#include "Proof.hpp"
+
+#include <boost/static_assert.hpp>
+
+#include "crypto/common.h"
+#include "libsnark/common/default_types/r1cs_ppzksnark_pp.hpp"
+#include "libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+
+using namespace libsnark;
+
+typedef default_r1cs_ppzksnark_pp curve_pp;
+typedef default_r1cs_ppzksnark_pp::G1_type curve_G1;
+typedef default_r1cs_ppzksnark_pp::G2_type curve_G2;
+typedef default_r1cs_ppzksnark_pp::GT_type curve_GT;
+typedef default_r1cs_ppzksnark_pp::Fp_type curve_Fr;
+typedef default_r1cs_ppzksnark_pp::Fq_type curve_Fq;
+typedef default_r1cs_ppzksnark_pp::Fqe_type curve_Fq2;
+
+namespace libzcash {
+
+BOOST_STATIC_ASSERT(sizeof(mp_limb_t) == 8);
+
+template<>
+Fq::Fq(curve_Fq element) : data()
+{
+    auto element_bigint = element.as_bigint();
+
+    unsigned char *p = data.begin();
+    for (ssize_t i = 3; i >= 0; i--, p += 8) {
+        WriteBE64(p, element_bigint.data[i]);
+    }
+}
+
+template<>
+curve_Fq Fq::to_libsnark_fq() const
+{
+    bigint<4> element_bigint;
+
+    const unsigned char *p = data.begin();
+    for (ssize_t i = 3; i >= 0; i--, p += 8) {
+        element_bigint.data[i] = ReadBE64(p);
+    }
+
+    // Check that the integer is smaller than the modulus
+    auto modq = curve_Fq::field_char();
+
+    for (ssize_t i = 3; i >= 0; i--) {
+        if (element_bigint.data[i] < modq.data[i]) {
+            return curve_Fq(element_bigint);
+        } else if (element_bigint.data[i] > modq.data[i]) {
+            throw std::logic_error("element is not in Fq");
+        }
+    }
+
+    throw std::logic_error("element is not in Fq");
+}
+
+template<>
+CompressedG1::CompressedG1(curve_G1 point)
+{
+    if (point.is_zero()) {
+        throw std::ios_base::failure("curve point is zero");
+    }
+
+    point.to_affine_coordinates();
+
+    x = Fq(point.X);
+    y_lsb = point.Y.as_bigint().data[0] & 1;
+}
+
+template<>
+curve_G1 CompressedG1::to_libsnark_g1() const
+{
+    curve_Fq x_coordinate = x.to_libsnark_fq<curve_Fq>();
+
+    // TODO: use the `b` from libsnark instead of hardcoding it here
+    // y = +/- sqrt(x^3 + b)
+    auto y = ((x_coordinate.squared() * x_coordinate) + curve_Fq("3")).sqrt();
+
+    if ((y.as_bigint().data[0] & 1) != y_lsb) {
+        y = -y;
+    }
+
+    curve_G1 r = curve_G1::one();
+    r.X = x_coordinate;
+    r.Y = y;
+    r.Z = curve_Fq::one();
+
+    if (!r.is_well_formed()) {
+        throw std::logic_error("point is not on the curve (G1)");
+    }
+
+    return r;
+}
+
+template<>
+CompressedG2::CompressedG2(curve_G2 point)
+{
+    if (point.is_zero()) {
+        throw std::ios_base::failure("curve point is zero");
+    }
+
+    point.to_affine_coordinates();
+
+    x0 = Fq(point.X.c0);
+    x1 = Fq(point.X.c1);
+    y_lsb = point.Y.c0.as_bigint().data[0] & 1;
```

Comment:
This should be using SORT compression, not LSB compression (which doesn't work for extension fields).